src/backend/metal/src/device.rs

```diff
@@ -1050,8 +1050,10 @@ impl hal::Device<Backend> for Device {
 
         // temporary command buffer to copy the contents from
         // the given buffers into the allocated CPU-visible buffers
-        let mut pool = self.command_shared.queue_pool.lock().unwrap();
-        let (queue_id, cmd_buffer) = pool.make_command_buffer(&self.device);
+        let (queue_id, cmd_buffer) = self.command_shared.queue_pool
```

Comment:
Ouch, is there a way we can improve this to avoid being bitten by this mutex again? It seems like it would be really easy to make this mistake – basically anytime we keep a lock active while calling other functions (which may or may not happen to try to acquire a lock somewhere down their callstack).

Maybe (internally) we could pass `queue_pool` through some parts call stack explicitly, so the caller can decide how the locking should work.

Reply:
> Maybe (internally) we could pass queue_pool through some parts call stack explicitly, so the caller can decide how the locking should work.

Could you elaborate on that?

I was thinking more in line of separating acquisition from release. The latter can just be `Arc<AtomicUint>`, which will be flushed on acquisition.